packages/relay-compiler/transforms/RelayGenerateIDFieldTransform.js

```diff
@@ -11,104 +11,113 @@
 
 'use strict';
 
-const {hasUnaliasedSelection} = require('./RelayTransformUtils');
+const invariant = require('invariant');
+const {hasSelection} = require('./RelayTransformUtils');
 const {
   assertAbstractType,
   assertCompositeType,
-  assertLeafType,
+  isInterfaceType,
+  getNullableType,
+  GraphQLInterfaceType,
+  GraphQLObjectType,
+  GraphQLSchema,
 } = require('graphql');
 const {
   CompilerContext,
   SchemaUtils,
   IRTransformer,
 } = require('graphql-compiler');
 
-import type {InlineFragment, LinkedField, ScalarField} from 'graphql-compiler';
-import type {GraphQLCompositeType} from 'graphql';
+import type {
+  Fragment,
+  InlineFragment,
+  LinkedField,
+  ScalarField,
+} from 'graphql-compiler';
+import type {GraphQLCompositeType, GraphQLField} from 'graphql';
+
 const {
   canHaveSelections,
   getRawType,
-  hasID,
   implementsInterface,
   isAbstractType,
   mayImplement,
 } = SchemaUtils;
 
 const ID = 'id';
+const ID_KEY = '__id';
 const ID_TYPE = 'ID';
 const NODE_TYPE = 'Node';
 
-type State = {
-  idField: ScalarField,
-};
-
 /**
- * A transform that adds an `id` field on any type that has an id field but
- * where there is no unaliased `id` selection.
+ * A transform that adds a `__id` field on any type that has a `Node` or `id`
+ * field but where there is no unaliased `__id` selection.
```

Comment:
so we'll always add an `__id` selection, *unless* there's already an *unaliased* `__id` selection.

what happens when there's already an aliased `__id` selection?, will we add another `__id` selection too?